crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -292,95 +283,342 @@ where
+            let pre_balance: Amount = self
+                .ctx
+                .read_pre(&counterpart_balance_key)?
+                .unwrap_or_default();
+            let post_balance: Amount = self
+                .ctx
+                .read_post(&counterpart_balance_key)?
+                .unwrap_or_default();
+            // Public keys must be the hash of the sources/targets
+            let address_hash = <[u8; 20]>::from(ripemd::Ripemd160::digest(
+                sha2::Sha256::digest(
+                    &counterpart.to_address_ref().serialize_to_vec(),
                 ),
+            ));
+            let mut diff = match post_balance.checked_sub(pre_balance) {
+                Some(diff) => SignedAmount::Positive(diff),
+                None => SignedAmount::Negative(pre_balance - post_balance),
             };
 
-        Ok(TransparentTransferData {
-            source,
-            target,
-            token,
-            amount,
-        })
+            if let ShieldedActionOwner::Minted = counterpart {
+                // When receiving ibc transfers we mint and also shield so we
+                // have two credits, we need to mock the mint balance as a
+                // negative change even if it is positive
+                if let SignedAmount::Positive(amt) = diff {
+                    diff = SignedAmount::Negative(amt);
+                }
+            }
+
+            result
+                .other
+                .entry(token)
+                .or_insert(BTreeMap::default())
+                .insert(address_hash, diff);
+        }
+
+        Ok(result)
+    }
+
+    fn validate_transparent_bundle(
+        &self,
+        shielded_tx: &Transaction,
+        changed_balances: ChangedBalances,
+        transparent_tx_pool: &mut I128Sum,
+        epoch: Epoch,
+        conversion_state: &ConversionState,
+    ) -> Result<bool> {
+        if let Some(transp_bundle) = shielded_tx.transparent_bundle() {
+            // Support structs to allow a valid check on the asset derivation.
+            // These also allow to check if some transparent inputs or outputs
+            // were not utilized for a given token which would lead to an
+            // inconsistent transparent balance of masp
+            let mut unprocessed_vins =
+                BTreeSet::from_iter(0..transp_bundle.vin.len());
+            let mut unprocessed_vouts =
+                BTreeSet::from_iter(0..transp_bundle.vout.len());
+
+            // Run the checks fore every token that involved the masp balances
+            for token in changed_balances.masp {
+                let denom = read_denom(&self.ctx.pre(), token)?.ok_or_err_msg(
+                    "No denomination found in storage for the given token",
+                )?;
+
+                let mut token_bundle_balances = BTreeMap::default();
+
+                // To help recognize asset types not in the conversion tree
+                let unepoched_tokens = unepoched_tokens(token, denom)?;
```

Comment:
Actually we still depend on the `token` binding in two more places inside this loop:

1. When checking the conversion
2. When matching the transparent bundle agains the changed keys in storage

For the second point we could simply iterate on the entire `token_bundle_balances` but for the first point I suspect we'd need to iterate anyway, so we'll end up just moving the for loop to a different place instead of removing it